TMEDIA-834 full screen carousel (off of tmedia 689)

JackHowa opened the pull request:
## Ticket

- [TMEDIA-834](https://arcpublishing.atlassian.net/browse/TMEDIA-834) subtask of [TMEDIA-689](https://arcpublishing.atlassian.net/browse/TMEDIA-689)

## Description

Use full screen for carousel. 

Note: I could not mock requestFullscreen on the the target element without potentially introducing bugs into our component itself. Let me know if you have suggestions. I was able to mock on the document but didn't on individual tests. 

Also note: this has limited support on safari ios. It works on ipad but not iphones safari. 

## Acceptance Criteria

- Hide the full screen button if browser api doesn't support it 
- Is accessible. Includes aria-label and aria-controls like existing buttons in this component 
- Allows enough customization and configuration 

## Test Steps

1. Checkout branch - `git checkout TMEDIA-834-toggle-full-screen`
2. Run Storybook `npm run storybook`
3. Go to a chromium browser with the carousel storybook http://localhost:60003/?path=/story/components-carousel--show-full-screen-toggle
4. Click "Full Screen" default button
5. See the carousel go full-screen
6. Click "Minimize Screen" default button to exit full screen
7. Click "Full Screen" default button again
8. Now exit full-screen by pressing escape on a mac or `f11` on pc https://developer.mozilla.org/en-US/docs/Web/API/Fullscreen_API/Guide#things_your_users_want_to_know
9. Go to another example utilizing custom enter and exit full-screen buttons http://localhost:60003/?path=/story/components-carousel--show-custom-full-screen-toggle

<img width="1437" alt="Screen Shot 2022-04-15 at 09 44 19" src="https://user-images.githubusercontent.com/5950956/163584827-8d3f7a81-f2c3-4dc3-8f59-9669a27c35a8.png">
<img width="729" alt="Screen Shot 2022-04-15 at 09 44 30" src="https://user-images.githubusercontent.com/5950956/163584826-9764fae7-6d50-46c0-adab-c3747aa872c1.png">


10. Repeat open and close 
11.  Go to carousel with multiple images http://localhost:60003/?path=/story/components-carousel--show-custom-full-screen-toggle-with-multiple-slides
12. Open the carousel
13. Now ensure that you can still go to next and previous photos 
14. Notice the classes include the `c-carousel__button` class as well as corresponding modifiers for open and close full-screen mode classes c-carousel__button c-carousel__button--enter-full-screen
15. Repeat steps of opening and closing on safari on mac
16. Repeat steps of opening and closing on safari for ipad. I used a simulator. (You'll notice that in the top-left corner there's an "x" that can't be removed according to the spec.) 
![image](https://user-images.githubusercontent.com/5950956/163261910-d0ff42d5-f297-4faf-ac2a-f412d333db6d.png)
17. You can test if this shows on safari, for instance, by removing the safari webkit check for rendering the buttons. Then running in safari. You won't see the button if you modify 
```js
	const fullScreenEnabledAllowed =
		(document.fullscreenEnabled || document.webkitFullscreenEnabled) && enableFullScreen;

```
to 
```js
	const fullScreenEnabledAllowed =
		(document.fullscreenEnabled /* ||  document.webkitFullscreenEnabled */) && enableFullScreen;
```
## Author Checklist

_The author of the PR should fill out the following sections to ensure this PR is ready for review._

- [x] Confirmed all the test steps a reviewer will follow above are working.
- [x] Ran this code locally and checked that there are not any unintended side effects. For example, that a CSS selector is scoped only to a particular block.
- [x] Confirmed relevant documentation has been updated/added.
- [x] Add label - **ready for review** when the pull request is ready for someone to begin reviewing

## Reviewer Checklist

_The reviewer of the PR should copy-paste this template into the review comments on review._

- [ ] All GitHub Actions pass
- [ ] Ran the code locally based on the test instructions.
- [ ] Checked Chromatic for Storybook changes, accepted the updates if acceptable
- [ ] Looked to see that the new or changed code has code coverage, specifically. We want the global code coverage to keep on going up with targeted testing.
- [ ] Approve and Add label - **ready to merge** if you are happy with the pull request
- [ ] Want another reviewer? Add the label **additional review**

----

## Questions 

- Is providing the custom button props and enable ok for the usecases? -> yes, follows existing pattern in component
- Is the behavior for showing and hiding full-screen what we want? You can test this on chromium or blink(safari) browsers.  -> Product says this is ok.
- Is the accessibility of this as expected? I wasn't sure if we needed/wanted an aria label for toggling screen; seemed unnecessary. -> There still are accessible reasons for this to exist. Updated based on best practices within this repo.

## Todos 

- [ ] Add event handler tests for the button -> can't really do this without hacking the component

## Notes 

- I'm splitting this ticket up into sub-tasks for clarity on my end https://arcpublishing.atlassian.net/jira/software/c/projects/TMEDIA/boards/875?modal=detail&selectedIssue=TMEDIA-689&assignee=5e387d6a1b1d910e5dfd7a9b
- Can be tested with `npm run storybook` and going to carousel custom full screen and default

JackHowa committed:
feat: Add show and exit full-screen functionality to the button

JackHowa committed:
fix: Add back accidental reverted export

nschubach commented on src/components/carousel/index.jsx:
```diff
@@ -70,6 +75,29 @@ const Carousel = ({
+			}
+		}
+
+		// safari needs prefix
+		if (document.webkitFullscreenEnabled) {
+			if (!document.webkitFullscreenElement) {
+				fullScreenElement.webkitRequestFullscreen().then(() => setIsFullScreen(true));
+			} else {
+				document.webkitExitFullscreen().then(() => setIsFullScreen(false));
+			}
+		}
+	};
```
Should put this in the else of the `if (document.fullscreenEnabled) {` I think.  It might be weird if both were supported by the browser for some reason.

JackHowa replied:
hmm gotcha. yeah I'll look at that. was working as expected in safari still

JackHowa replied:
updated @nschubach

nschubach commented on src/components/carousel/index.jsx:
```diff
@@ -108,6 +136,36 @@ const Carousel = ({
+			className={`${COMPONENT_CLASS_NAME}__full-screen-toggle`}
+		>
+			Full Screen
+		</button>
```
Wondering if we shouldn't mimic the prev/next buttons here.  If we use the standard Button component then it can be styled in a consistent manner for each state.  Currently, it would be styled differently than the prev/next for this specific button.

Thoughts?

JackHowa replied:
Yeah I can look at modelling off of that implementation

JackHowa replied:
updated @nschubach

nschubach commented on src/components/carousel/index.jsx:
```diff
@@ -37,10 +37,15 @@ const Carousel = ({
 	slidesToShow,
+	fullScreenShowButton,
+	fullScreenMinimizeButton,
+	enableFullScreenToggleButton,
```
Not sold on my thoughts here, but `enableFullScreenToggleButton` feels verbose.  I think I would prefer to see `enableFullScreen` since it's enabling the functionality, not just the button.

JackHowa replied:
gotcha. yeah I can change it if that feels weird. I mostly didn't use `enableFullScreen` because I didn't want people to think that that was going to be full-screen on first render if it was enabled. maybe `enableFullScreenOption`?

nschubach replied:
`enableFullScreenOption` works.

I think if the variable was `fullScreen` that might indicate that it's full screen at start, but I can settle on the `enableFullScreenOption`.  I'm just thinking about it as a tag option like so:
```
  <Carousel fullScreen /> /* this feels very much like fullScreen is active */
  <Carousel enableFullScreen />  /* this feels like it's enabling the fullScreen feature to me */
  <Carousel enableFullScreenOption /> /* this feels alright to me though, I'm ok with this little verbosity */
  <Carousel enableFullScreenToggleButton /> /* this just feels like it's enabling the button, as odd as that sounds */
```

JackHowa replied:
it's ok. @nschubach I went with your original suggestion! thought about it some more and it's cool https://github.com/WPMedia/arc-themes-components/pull/64/commits/c4367a200c4d1d9046a89ae341751b1fd4a1379c

It's more succinct. And "enabling" still implies an option. I have a tendency toward verbosity in variable naming

JackHowa committed:
refactor: Use safari with else per comment to ensure webkit and fullscreen enabled don't interefere

Per https://github.com/WPMedia/arc-themes-components/pull/64#discussion_r849565955

JackHowa committed:
Merge branch 'arc-themes-release-version-2.00' into TMEDIA-834-toggle-full-screen

JackHowa committed:
fix: Only show buttons for full-screen if that api is allowed on browser

Per comment https://arcpublishing.atlassian.net/browse/TMEDIA-834?focusedCommentId=734128

JackHowa committed:
refactor: Update name of enable prop for full screen

Per https://github.com/WPMedia/arc-themes-components/pull/64#discussion_r849587982

JackHowa committed:
feat: Use aria label and aria control with the id to point to the carousel, reusing child button component pattern

Per https://github.com/WPMedia/arc-themes-components/pull/64#discussion_r849581154

JackHowa committed:
feat: Use same naming standard for buttons. Ensure buttons follow same style pattern as next and previous buttons

JackHowa committed:
test: Remove event handler test coverage support as it can't really be mocked well

JackHowa commented:
@nschubach addressed your suggestions

matthewroach commented on src/components/carousel/index.jsx:
```diff
@@ -129,6 +204,13 @@ const Carousel = ({
 
 			<div className="c-carousel__actions">
+				{/* only show button at all if enabled on the document */}
+				{fullScreenEnabledAllowed && !isFullScreen ? resolvedFullScreenShowButton : null}
```
The full screen button should be placed at the top of the carousel - adding them into this actions container isn't the correct choice and will cause issues with styling and tab order.
We will want to have a container that sits at the top of the carousel to house this feature, auto play and the counter

JackHowa replied:
ok thanks, updated @matthewroach

JackHowa committed:
feat: Set top actions container to container full-screen button

Per feedback https://github.com/WPMedia/arc-themes-components/pull/64#discussion_r849893125

JackHowa committed:
feat: Mimic bottom actions container styles for container layout

JackHowa committed:
Merge branch 'arc-themes-release-version-2.00' into TMEDIA-834-toggle-full-screen

JackHowa committed:
feat: Update to use grid for showing actions above the picture rather than atop

nschubach reviewed (approved):
Everything seems to be in order!
